File: web/js/MotorBloques.js
class Bloque {
    constructor(x, y, ancho, alto, tipo) {
        this.x = x;
        this.y = y;
        this.ancho = ancho;
        this.alto = alto;
        this.tipo = tipo; // Ej: '300', '400', '500'
    }
}

class MotorBloques {
    constructor(config) {
        this.config = config;
        this.bloquesColocados = [];
    }

    /**
     * Motor principal: Genera la disposición de bloques fila por fila.
     */
    run() {
        this.bloquesColocados = [];
        const { anchoArea, altoArea, junta, tamanosBloque, offsetMataJunta, altoBloque } = this.config;
        
        // Convertir array de tamaños a números enteros y ordenar de mayor a menor
        const ANCHOS_DISPONIBLES = tamanosBloque.map(Number).sort((a, b) => b - a);
        const ANCHO_MINIMO_REQUERIDO = ANCHOS_DISPONIBLES[ANCHOS_DISPONIBLES.length - 1] + junta;
        
        let yActual = 0; // Coordenada Y de la fila actual

        // Iterar fila por fila hasta exceder la altura del área
        let numFila = 0;
        while (yActual + altoBloque <= altoArea) { 
            
            let xActual = 0; // Coordenada X (inicio) de la fila
            
            // Implementar Mata-Junta (Offset)
            const offset = (numFila % 2 !== 0) ? offsetMataJunta : 0;
            xActual += offset;
            
            // Si el espacio restante es menor que el bloque más pequeño, terminamos
            if (anchoArea - xActual < ANCHO_MINIMO_REQUERIDO) {
                // Si la fila tiene espacio para el bloque más chico, pero no para un bloque completo, NO colocamos nada
                if (numFila === 0 || bloquesFila.length === 0) break; 
            }

            let bloquesFila = [];

            // Algoritmo: Llenar la fila
            while (xActual < anchoArea) { 
                
                // Calculamos cuánto espacio queda *hasta el final* del muro
                const espacioRestanteAlMuro = anchoArea - xActual;

                // 1. PRIMERA BÚSQUEDA: Encontrar ajuste perfecto (Regla No-Corte)
                let mejorBloqueNoCorte = null;
                let mejorSobrante = this.config.tolerancia + 1;

                for (const anchoBloque of ANCHOS_DISPONIBLES) {
                    const anchoTotalBloque = anchoBloque + junta;
                    const sobrante = espacioRestanteAlMuro - anchoTotalBloque; 

                    // Condición No-Corte: El sobrante debe ser positivo y caer dentro de la tolerancia
                    if (sobrante >= 0 && sobrante <= this.config.tolerancia) {
                        if (sobrante < mejorSobrante) {
                            mejorBloqueNoCorte = anchoBloque;
                            mejorSobrante = sobrante;
                            if (sobrante === 0) break; // Ajuste perfecto!
                        }
                    }
                }
                
                // 2. COLOCACIÓN
                if (mejorBloqueNoCorte) {
                    // Caso A: Se encontró un ajuste perfecto/tolerable (No-Corte)
                    const anchoTotal = mejorBloqueNoCorte + junta;
                    
                    bloquesFila.push(new Bloque(
                        xActual,
                        yActual,
                        mejorBloqueNoCorte,
                        altoBloque,
                        mejorBloqueNoCorte.toString()
                    ));
                    
                    // Mueve X al final del muro (salto) para terminar la fila
                    xActual = anchoArea; 
                    
                } else {
                    // Caso B: No se encontró ajuste perfecto. Usamos Greedy para seguir llenando el espacio restante.
                    let anchoGreedy = null;

                    // Buscamos el bloque MÁS GRANDE que cabe en el espacio restante.
                    for (const anchoBloque of ANCHOS_DISPONIBLES) {
                         const anchoTotalBloque = anchoBloque + junta;
                         
                         // Sólo colocamos si el bloque cabe Y si después de colocarlo, aún queda espacio
                         // o si el espacio restante es mayor que el bloque más chico (para evitar dejar huecos grandes)
                         if (xActual + anchoTotalBloque <= anchoArea) {
                            // Se encontró el más grande que cabe.
                            anchoGreedy = anchoBloque;
                            break; 
                         }
                    }

                    if (anchoGreedy) {
                        const anchoTotal = anchoGreedy + junta;
                        bloquesFila.push(new Bloque(
                            xActual, 
                            yActual, 
                            anchoGreedy, 
                            altoBloque, 
                            anchoGreedy.toString()
                        ));
                        xActual += anchoTotal;
                    } else {
                        // El espacio restante es demasiado pequeño para colocar incluso el bloque más chico.
                        break; 
                    }
                }
            }
            
            // Guardar los bloques de la fila completa
            this.bloquesColocados.push(...bloquesFila);

            // Mover a la siguiente fila
            yActual += altoBloque + junta;
            numFila++;
        }

        return this.bloquesColocados;
    }
}

File: Motor/MotorBloques.cs
using System;
using System.Collections.Generic;
using System.Linq;
using MotorBloques.Models; // Necesario para usar Bloque y Configuracion

namespace MotorBloques.Motor
{
    public class MotorBloques
    {
        private readonly Configuracion _config;
        private List<Bloque> _bloquesColocados;

        // -----------------------------------------------------------------
        // CONSTRUCTOR CORREGIDO (Llama a la conversión de unidades)
        // -----------------------------------------------------------------
        public MotorBloques(Configuracion config)
        {
            // Convertir inmediatamente todos los valores a MM
            _config = ConvertToInternalUnits(config); 
            _bloquesColocados = new List<Bloque>();
        }

        // -----------------------------------------------------------------
        // LÓGICA DE CONVERSIÓN DE UNIDADES (Tarea de la conversación con René)
        // -----------------------------------------------------------------
        
        /// <summary>
        /// Obtiene el factor de multiplicación necesario para convertir la unidad de entrada a milímetros (mm).
        /// </summary>
        private double GetConversionFactor(string unidad)
        {
            return unidad.ToLowerInvariant() switch
            {
                "m" => 1000.0, 
                "cm" => 10.0,  
                "mm" => 1.0,   
                _ => 1.0,      // Si no se reconoce, asume milímetros.
            };
        }

        /// <summary>
        /// Convierte todas las dimensiones de la configuración de entrada a milímetros.
        /// </summary>
        private Configuracion ConvertToInternalUnits(Configuracion originalConfig)
        {
            var factor = GetConversionFactor(originalConfig.UnidadEntrada);
            if (factor == 1.0) return originalConfig;

            var convertedConfig = new Configuracion
            {
                // Conversión de propiedades individuales
                AnchoArea = (int)(originalConfig.AnchoArea * factor),
                AltoArea = (int)(originalConfig.AltoArea * factor),
                Junta = (int)(originalConfig.Junta * factor),
                Tolerancia = (int)(originalConfig.Tolerancia * factor),
                OffsetMataJunta = (int)(originalConfig.OffsetMataJunta * factor),
                AltoBloque = (int)(originalConfig.AltoBloque * factor),
                GrosorParedFinal = (int)(originalConfig.GrosorParedFinal * factor),
                
                // Conversión de la lista de tamaños
                TamanosBloque = originalConfig.TamanosBloque.Select(t => (int)(t * factor)).ToList(),
                UnidadEntrada = "mm" // La nueva unidad interna (estándar)
            };
            return convertedConfig;
        }


        // -----------------------------------------------------------------
        // LÓGICA PRINCIPAL DEL MOTOR (Método Run)
        // -----------------------------------------------------------------
        
        /// <summary>
        /// Genera la disposición de bloques fila por fila, aplicando las reglas de No-Corte y Mata-Junta.
        /// </summary>
        public List<Bloque> Run()
        {
            _bloquesColocados.Clear();

            // Desestructurar configuración (ya en MM)
            int anchoArea = _config.AnchoArea;
            int altoArea = _config.AltoArea;
            int junta = _config.Junta;
            int tolerancia = _config.Tolerancia;
            int altoBloque = _config.AltoBloque;
            int offsetMataJunta = _config.OffsetMataJunta;

            // Ordenar la lista para la lógica Greedy
            var anchosDisponibles = _config.TamanosBloque.OrderByDescending(a => a).ToArray();
            if (!anchosDisponibles.Any()) return _bloquesColocados;

            int anchoMinimoRequerido = anchosDisponibles.Last() + junta;

            int yActual = 0; // Coordenada Y de la fila actual
            int numFila = 0;
            
            // Bucle principal para iterar por filas
            while (yActual + altoBloque <= altoArea)
            {
                int xActual = 0; // Coordenada X (inicio) de la fila

                // 1. Implementar Mata-Junta (Offset)
                int offset = (numFila % 2 != 0) ? offsetMataJunta : 0;
                xActual += offset;
                
                // Si el espacio restante es demasiado pequeño para el bloque más chico, terminamos.
                if (anchoArea - xActual < anchoMinimoRequerido) break;

                var bloquesFila = new List<Bloque>();

                // Bucle para llenar la fila
                while (xActual < anchoArea)
                {
                    double espacioRestanteAlMuro = anchoArea - xActual;

                    // 2. PRIMERA BÚSQUEDA: Encontrar ajuste perfecto (Regla No-Corte)
                    int? mejorBloqueNoCorte = null;
                    int mejorSobrante = tolerancia + 1; 

                    foreach (var anchoBloque in anchosDisponibles)
                    {
                        int anchoTotalBloque = anchoBloque + junta;
                        int sobrante = (int)(espacioRestanteAlMuro - anchoTotalBloque);

                        // Condición No-Corte
                        if (sobrante >= 0 && sobrante <= tolerancia)
                        {
                            if (sobrante < mejorSobrante)
                            {
                                mejorBloqueNoCorte = anchoBloque;
                                mejorSobrante = sobrante;
                                if (sobrante == 0) break;
                            }
                        }
                    }

                    // 3. COLOCACIÓN
                    if (mejorBloqueNoCorte.HasValue)
                    {
                        // Caso A: Ajuste perfecto/tolerable
                        bloquesFila.Add(new Bloque(
                            xActual, 
                            yActual, 
                            mejorBloqueNoCorte.Value, 
                            altoBloque, 
                            mejorBloqueNoCorte.Value.ToString()
                        ));

                        xActual = anchoArea; // Mueve X al final del muro
                    }
                    else
                    {
                        // Caso B: Greedy para seguir llenando
                        int? anchoGreedy = null;

                        // Buscamos el bloque MÁS GRANDE que cabe
                        foreach (var anchoBloque in anchosDisponibles)
                        {
                            int anchoTotalBloque = anchoBloque + junta;
                            if (xActual + anchoTotalBloque <= anchoArea)
                            {
                                anchoGreedy = anchoBloque;
                                break; 
                            }
                        }

                        if (anchoGreedy.HasValue)
                        {
                            int anchoTotal = anchoGreedy.Value + junta;
                            bloquesFila.Add(new Bloque(
                                xActual, 
                                yActual, 
                                anchoGreedy.Value, 
                                altoBloque, 
                                anchoGreedy.Value.ToString()
                            ));
                            xActual += anchoTotal;
                        }
                        else
                        {
                            // Espacio restante demasiado pequeño
                            break;
                        }
                    }
                } // Fin del bucle de llenado de fila

                _bloquesColocados.AddRange(bloquesFila);
                yActual += altoBloque + junta;
                numFila++;
            }

            return _bloquesColocados;
        }
    }
}